torchtune/utils/_checkpointing/_checkpointer_utils.py

```diff
@@ -15,6 +15,101 @@ class ModelType(Enum):
+    A bare-bones class meant for checkpoint save and load for optimizers running
+    in backward. Usage is limited to the following:
+
+    optim_dict = {
+        p: config.instantiate(cfg_optimizer, [p])
+        for p in self._model.parameters()
+    }
+    # Save checkpoint
+    ckpt = OptimizerInBackwardWrapper(optim_dict).state_dict()
+    torch.save("/tmp/optim_ckpt", ckpt)
+    # Load checkpoint
+    placeholder_optim_dict = {
+        p: config.instantiate(cfg_optimizer, [p])
+        for p in self._model.parameters()
+    }
+    wrapper = OptimInBackwardWrapper(placeholder_optim_dict)
+    # load_state_dict expects a dict produced by this class's
+    # state_dict method.
+    wrapper.load_state_dict(torch.load("/tmp/optim_ckpt"))
+    # placeholder_optim_dict now has updated optimizer states.
+
+    NOTE: This wrapper is only meant to be used for single-device use cases.
+        Distributed use cases such as FSDP, which require specialized
+        optimizer state checkpointing, are not supported.
+
+    """
+
+    def __init__(self, optim_map: Dict[str, torch.optim.Optimizer]):
+        self.optim_map = optim_map
+
+    def state_dict(self):
+        return {p: opt.state_dict() for p, opt in self.optim_map.items()}
+
+    def load_state_dict(self, optim_ckpt_map: Dict[str, Any]):
+        params_covered = set()
+        for param_name in optim_ckpt_map.keys():
+            if param_name not in self.optim_map:
+                raise RuntimeError(
+                    f"Trying to load optimizer state for unexpected param {param_name}"
+                )
+            self.optim_map[param_name].load_state_dict(optim_ckpt_map[param_name])
+            params_covered.add(param_name)
+        # Ensure all params have been loaded into, report missing params
+        missing_params = set(self.optim_map.keys()) - params_covered
+        if missing_params:
+            raise RuntimeError(
+                f"Expected to load optimizer state for params {missing_params}!"
+            )
+
+
+def create_optim_in_bwd_wrapper(
+    model: torch.nn.Module, optim_dict: Dict[torch.nn.Parameter, torch.optim.Optimizer]
+) -> OptimizerInBackwardWrapper:
+    """
+    Create a wrapper for optimizer step running in backward.
+
+    Args:
+        model (torch.nn.Module): Model that contains parameters that are being optimized. For now,
+            it is assumed that all parameters being optimized belong to a single top-level model.
+            `named_parameters` attribute of `model` will be accessed to look up parameter names for
+            parameters being optimized.
+        optim_dict (Dict[torch.nn.Parameter, torch.optim.Optimizer]): Mapping from
+            parameters to optimizers.
+
+    Returns:
+        ``OptimizerInBackwardWrapper``: Wrapper for optimizer states running in backward.
+    """
+    return OptimizerInBackwardWrapper(
+        {n: optim_dict[p] for n, p in model.named_parameters()}
+    )
+
+
+def register_optim_in_bwd_hooks(
```

Comment:
Probably move this stuff to an optim_utils at this point

Reply:
For now can also put in utils/_memory.py? If my interpretation of that file is correct I guess it includes memory-saving techniques given that AC wrapper is in there

Reply:
Yeah can do that also. My vote for memory saving utils would be to put them in a directory for each components, i.e. utils.act_ckpt, utils.optim, etc. But happy to go with this approach as well.